ruma-events-macros/src/event_enum.rs

```diff
@@ -575,17 +568,109 @@ fn field_return_type(name: &str) -> TokenStream {
 
 /// Custom keywords for the `event_enum!` macro
 mod kw {
-    syn::custom_keyword!(name);
+    syn::custom_keyword!(kind);
     syn::custom_keyword!(events);
 }
 
+pub enum EnumKind {
+    Basic(Ident),
+    Ephemeral(Ident),
+    Message(Ident),
+    State(Ident),
+    ToDevice(Ident),
+}
+
+impl EnumKind {
+    /// Return the span of this ident.
+    fn span(&self) -> Span {
+        match &self {
+            Self::Basic(i)
+            | Self::Ephemeral(i)
+            | Self::Message(i)
+            | Self::State(i)
+            | Self::ToDevice(i) => i.span(),
+        }
+    }
+    /// `[Kind]Event`
+    fn to_full_event_struct(&self) -> Ident {
+        match &self {
+            Self::Basic(i)
+            | Self::Ephemeral(i)
+            | Self::Message(i)
+            | Self::State(i)
+            | Self::ToDevice(i) => format_ident!("{}Event", i),
+        }
+    }
+
+    /// `Any[Kind]Event`
+    fn to_full_event_enum(&self) -> Ident {
+        match self {
+            Self::Basic(i)
+            | Self::Ephemeral(i)
+            | Self::Message(i)
+            | Self::State(i)
+            | Self::ToDevice(i) => format_ident!("Any{}Event", i),
+        }
+    }
+
+    /// `Any[Kind]EventStub`
+    fn to_event_stub_enum(&self) -> Option<Ident> {
+        match self {
+            Self::Ephemeral(i) | Self::Message(i) | Self::State(i) => {
+                Some(format_ident!("Any{}EventStub", i))
+            }
+            Self::Basic(_) | Self::ToDevice(_) => None,
+        }
+    }
+
+    /// `AnyStrippedStateEvent`
+    fn to_stripped_event_enum(&self) -> Option<Ident> {
+        match self {
+            Self::State(i) => Some(format_ident!("AnyStripped{}EventStub", i)),
+            Self::Ephemeral(_) | Self::Message(_) | Self::Basic(_) | Self::ToDevice(_) => None,
+        }
+    }
+
+    /// `Any[kind]EventContent`
+    fn to_content_enum(&self) -> Ident {
+        match self {
+            Self::Basic(i)
+            | Self::Ephemeral(i)
+            | Self::Message(i)
+            | Self::State(i)
+            | Self::ToDevice(i) => format_ident!("Any{}EventContent", i),
+        }
+    }
+}
+
+impl Parse for EnumKind {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        Ok(match input.parse::<Ident>()? {
+            i if i.to_string().as_str() == "Basic" => EnumKind::Basic(i),
+            i if i.to_string().as_str() == "EphemeralRoom" => EnumKind::Ephemeral(i),
+            i if i.to_string().as_str() == "Message" => EnumKind::Message(i),
+            i if i.to_string().as_str() == "State" => EnumKind::State(i),
+            i if i.to_string().as_str() == "ToDevice" => EnumKind::ToDevice(i),
```

Comment:
This is a bit repetitive. Why not bind the ident to a local variable before the match and then match on `ident.to_string().as_str()`?